proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+Furthermore, none of these libraries contain implementations of post-quantum cryptography schemes; although existing quantum computers are of insufficient capability to break commonly used algorithms such as Ed25519, this will not remain true forever, and it would be best to have quantum-resistant implementations ready for adoption long before such attacked become practical.
+
+Each of these efforts has one or more of these issues that preclude them from being considered a complete success:
+
+- Insufficient abstraction
+- Maintenance of handwritten C
+- Bindings to an insufficiently general library
+- Burdensome dependencies
+- Dead project
+- Lack of post-quantum algorithms
+
+However, there is also a lot to be learned from the ways these prior efforts *were* successful. This proposal will assess these existing efforts in an attempt to consolidate their learnings and functionality.
+
+## An example of a successful ecosystem
+
+The health of the Haskell server ecosystem may be used as an example of a successful community-driven niche, and we can aspire to bring that same quality and health to the Haskell cryptography ecosystem. There are many pertinent parallels in terms of managing performance, complexity, and safety that we can learn from:
+
+- Low-level packages like `wai` supporting multiple backends
+- Backends range from simple like `scotty` to complex like `servant`
+- Frontends like `blaze-html` and `lucid`
+- Libraries can take advantage of type system to provide good abstractions
+- Flexible enough abstractions to explore the problem space with multiple solutions
+- Appropriate licensing
+
+`servant` (and the `wai` ecosystem in general) is a success story and an example of higher-order programming making something safer and easier, something that we can aspire to do with cryptography.
+
+## Technical Content
+
+This proposal is for the contruction of a suite of cryptography libraries, ranging from low-level bindings to specific algorithms, to higher-level abstractions and typeclasses.
+
+This leg of the proposal focuses on bindings to the `botan` cryptography, in the form of the following libraries:
+
+### botan-bindings
+
+The `botan-bindings` library will contain raw bindings and is an almost direct, 1-1 translation of the C API into Haskell FFI calls. As such, it exposes and operates over C FFI types, requires buffer and pointer and pointer management, and returns error codes. This library only exposes FFI calls and constants, and is suitable for building your own abstraction over Botan.
+
+### botan-low
+
+The `botan-low` library will contain low-level bindings which wrap the FFI calls into monadic IO actions. This library handles the translation between buffers and ByteStrings, and throws exceptions in the case of errors, but is otherwise a fairly faithful translation of the interface. Thus this library is suitable for use, but is not recommended for day-to-day programming due to statefulness and a lack of type safety.
```

Comment:
How is it not type safe?

Reply:
It uses stringly-typed names (that follow a specific, undocumented parsable format) to pick between between algorithms with significantly different behavioral characteristics, the context / state is an opaque pointer, there is no way to tell what algorithm a context is for, and it relies on exceptional return codes when the user attempts to call a function with improper inputs. Quite frankly, I'd go as far as to say that `botan-low` is wildly unsafe unless you are excruciatingly careful, which is why `botan` is separate.